src/ws/resharder.ts

```diff
@@ -1,31 +1,102 @@
-import { GatewayManager } from "../bot.ts";
+import { createGatewayManager, GatewayManager } from "../bot.ts";
 import { GetGatewayBot } from "../types/gateway/getGatewayBot.ts";
 
 /** The handler to automatically reshard when necessary. */
-export async function resharder(gateway: GatewayManager) {
-  // TODO: is it possible to route this to REST?
-  const results = (await fetch(`https://discord.com/api/gateway/bot`, {
-    headers: { Authorization: gateway.token },
-  }).then((res) => res.json())) as GetGatewayBot;
+export async function resharder(oldGateway: GatewayManager, results: GetGatewayBot) {
+  oldGateway.debug("[Resharding] Starting the reshard process.");
 
-  const percentage = ((results.shards - gateway.maxShards) / gateway.maxShards) * 100;
-  // Less than necessary% being used so do nothing
-  if (percentage < gateway.reshardPercentage) return;
-
-  // Don't have enough identify rate limits to reshard
-  if (results.sessionStartLimit.remaining < results.shards) {
-    return;
-  }
+  const gateway = createGatewayManager({
+    ...oldGateway,
+    // IGNORE EVENTS FOR NOW
+    handleDiscordPayload: async function () {},
+  });
 
   // Begin resharding
   gateway.maxShards = results.shards;
+  // FOR MANUAL SHARD CONTROL, OVERRIDE THIS SHARD ID!
+  gateway.lastShardId = oldGateway.lastShardId === oldGateway.maxShards ? gateway.maxShards : oldGateway.lastShardId;
+  gateway.shardsRecommended = results.shards;
+  gateway.sessionStartLimitTotal = results.sessionStartLimit.total;
+  gateway.sessionStartLimitRemaining = results.sessionStartLimit.remaining;
+  gateway.sessionStartLimitResetAfter = results.sessionStartLimit.resetAfter;
+  gateway.maxConcurrency = results.sessionStartLimit.maxConcurrency;
   // If more than 100K servers, begin switching to 16x sharding
   if (gateway.maxShards && gateway.useOptimalLargeBotSharding) {
+    gateway.debug("[Resharding] Using optimal large bot sharding solution.");
     gateway.maxShards = Math.ceil(
       gateway.maxShards /
         (results.sessionStartLimit.maxConcurrency === 1 ? 16 : results.sessionStartLimit.maxConcurrency),
     );
   }
 
   gateway.spawnShards(gateway, gateway.firstShardId);
+
+  return new Promise((resolve) => {
+    // TIMER TO KEEP CHECKING WHEN ALL SHARDS HAVE RESHARDED
+    const timer = setInterval(async () => {
+      const pending = await gateway.resharderIsPending(gateway, oldGateway);
+      // STILL PENDING ON SOME SHARDS TO BE CREATED
+      if (pending) return;
+
+      // ENABLE EVENTS ON NEW SHARDS AND IGNORE EVENTS ON OLD
+      gateway.handleDiscordPayload = oldGateway.handleDiscordPayload;
+      oldGateway.handleDiscordPayload = function () {};
+
+      // STOP TIMER
+      clearInterval(timer);
+      await gateway.resharderCloseOldShards(oldGateway);
```

Comment:
The handler function gets removed from the old gateway. 
The shards are closed later whilst the closer waits for the queues to finish. The problem now is that if it was for example a request guild members request, then this will never finish since the handler function does not receive the response from discord anymore.
Has this been considered?